### PR TITLE
ref(integrations): Update delete flow and allow reinstalls for integrations

### PR DESCRIPTION
**Pull request by MeredithAnya**
This PR covers the following:

-  **Update link to differentiate between user and org accounts so we link to the correct place**
-  **Handle new installations – locate old installation and re-enable**
- **Add reinstall button that sends user back through flow**

screen shot 2018-06-26 at 1 05 56 pm

- **Provide better messaging for disabling/deleting integrations**

screen shot 2018-06-27 at 11 12 00 am

### CODE REVIEW

File: src/sentry/integrations/pipeline.py [hunk 1]
```diff
@@ -50,10 +51,27 @@ class IntegrationPipeline(Pipeline):
 
     def finish_pipeline(self):
         data = self.provider.build_integration(self.state.data)
+        try:
+            response = self.reinstall(data)
+        except Integration.DoesNotExist:
+            pass
+        else:
+            return response
```
Comment (EvanPurkhiser) [inline src/sentry/integrations/pipeline.py]: Can't this just be written

```py
try:
    return self.reinstall(data)
except Integration.DoesNotExist:
   pass
```

File: src/sentry/integrations/pipeline.py [hunk 2]
```diff
@@ -50,10 +51,27 @@ class IntegrationPipeline(Pipeline):
...
 
+    def reinstall(self, data):
+        integration = Integration.objects.get(
+            provider=self.provider.key,
+            name=data['name']
+        )
+        org_integration = OrganizationIntegration.objects.get(
+            integration=integration, organization=self.organization)
+        integration.update(external_id=data['external_id'], status=ObjectStatus.VISIBLE)
+        integration.reinstall_repositories()
```
Comment (EvanPurkhiser) [inline src/sentry/integrations/pipeline.py]: Can we rename this function to just `reinstall` and have it on the Integration service object (`get_install`) and then in the GH integration do the instal repo work or something along those lines.

The way this is right now just breaks abstraction a little bit, the integration pipeline probably shouldn't know anything about re-enabling repositories.

File: src/sentry/integrations/pipeline.py [hunk 3]
```diff
@@ -50,10 +51,27 @@ class IntegrationPipeline(Pipeline):
...
+            response = self.reinstall(data)
+        except Integration.DoesNotExist:
+            pass
+        else:
+            return response
         response = self._finish_pipeline(data)
         self.clear_session()
         return response
 
+    def reinstall(self, data):
```
Comment (EvanPurkhiser) [inline src/sentry/integrations/pipeline.py]: Do we actually need a reinstall method? The _finish_pipeline should be idempotent enough that running through it will be the equivalent of a reinstall.

We would need to set the `status` of the integration and then call the reinstall stuff if the organization integration wasn't created during the install

File: src/sentry/static/sentry/app/components/confirm.jsx [hunk 4]
```diff
@@ -15,6 +15,7 @@ class Confirm extends React.PureComponent {
      * If true, will skip the confirmation modal and call `onConfirm`
      */
     bypass: PropTypes.bool,
+    title: PropTypes.string,
```
Comment (EvanPurkhiser) [inline src/sentry/static/sentry/app/components/confirm.jsx]: Can you make this a separate PR? This looks like it's going to be break a lot of our confirm modals right away.

File: src/sentry/static/sentry/app/views/organizationIntegrations/addIntegrationButton.jsx [hunk 5]
```diff
@@ -31,6 +31,7 @@ export default class AddIntegrationButton extends React.Component {
   static propTypes = {
     provider: PropTypes.object.isRequired,
     onAddIntegration: PropTypes.func.isRequired,
+    reinstall: PropTypes.bool,
```
Comment (EvanPurkhiser) [inline src/sentry/static/sentry/app/views/organizationIntegrations/addIntegrationButton.jsx]: Let's have this take a `text` property instead, it will be a wee bit more generic, but then it will render the text if passed as a prop, otherwise it will render `buttonText`.

File: src/sentry/static/sentry/app/views/organizationIntegrations/installedIntegration.jsx [hunk 6]
```diff
@@ -37,13 +39,18 @@ export default class InstalledIntegration extends React.Component {
     );
   }
 
+  mergeIntegration() {}
```
Comment (EvanPurkhiser) [inline src/sentry/static/sentry/app/views/organizationIntegrations/installedIntegration.jsx]: Don't we have to do something to make it un-gray

File: src/sentry/static/sentry/app/views/organizationIntegrations/installedIntegration.jsx [hunk 7]
```diff
@@ -37,13 +39,18 @@ export default class InstalledIntegration extends React.Component {
...
+
   renderDisableIntegration(integration) {
-    const message = `You must uninstall this integration from
-      %s in order to delete this integration in Sentry.`;
+    let {body, action_text} = integration.provider.aspects.disable_dialog;
```
Comment (EvanPurkhiser) [inline src/sentry/static/sentry/app/views/organizationIntegrations/installedIntegration.jsx]: nit: I've been trying to use `const` for new components that aren't using `let`. Not trying to force it on you, but I think consistency within files at least is good.

File: src/sentry/static/sentry/app/views/organizationIntegrations/installedIntegration.jsx [hunk 8]
```diff
@@ -37,13 +39,18 @@ export default class InstalledIntegration extends React.Component {
...
+    let message = <Alert type="info" icon="icon-circle-exclamation">{body}</Alert>;
+
     return (
       <Confirm
-        confirmText={t('Remove on GitHub')}
-        message={t(message, integration.provider.key)}
+        confirmText={t(action_text)}
```
Comment (EvanPurkhiser) [inline src/sentry/static/sentry/app/views/organizationIntegrations/installedIntegration.jsx]: Can we do the translation in django actually? I think this is consistent with the rest of the metadata stuff? maybe I'm wrong

File: src/sentry/static/sentry/app/views/organizationIntegrations/installedIntegration.jsx (hunk 7 above)
Comment (EvanPurkhiser) [inline src/sentry/static/sentry/app/views/organizationIntegrations/installedIntegration.jsx]: We should probably make the aspects Javascript style camel case.

File: src/sentry/static/sentry/app/views/organizationIntegrations/installedIntegration.jsx [hunk 9]
```diff
@@ -52,15 +59,10 @@ export default class InstalledIntegration extends React.Component {
...
-      you want to remove this integration?`;
+    let {body, action_text} = integration.provider.aspects.removal_dialog;
+    let message = <Alert type="warning" icon="icon-circle-exclamation">{body}</Alert>;
     return (
-      <Confirm
-        disabled={integration.status === 'disabled'}
-        message={t(message)}
-        onConfirm={() => this.props.onRemove()}
-      >
+      <Confirm message={t(message)} confirmText={t(action_text)} priority="danger" title="Deleting Your Integration" onConfirm={() => this.props.onRemove()}>
```
Comment (EvanPurkhiser) [inline src/sentry/static/sentry/app/views/organizationIntegrations/installedIntegration.jsx]: I'm surprised prettier didn't multiline this 🤔

File: src/sentry/static/sentry/app/views/organizationIntegrations/installedIntegration.jsx [hunk 10]
```diff
@@ -77,16 +79,29 @@ export default class InstalledIntegration extends React.Component {
...
+                  {t('Configure')}
+                </Button>
+              </Box>
+            )}
+          {integration.status === 'disabled' &&
+            this.hasConfiguration() && (
+              <Box mr={1}>
+                <AddIntegrationButton
+                  size="small"
+                  priority="danger"
```
Comment (EvanPurkhiser) [inline src/sentry/static/sentry/app/views/organizationIntegrations/installedIntegration.jsx]: Should it really be priority danger?

File: src/sentry/static/sentry/app/views/organizationIntegrations/installedIntegration.jsx [hunk 11]
```diff
@@ -77,16 +79,29 @@ export default class InstalledIntegration extends React.Component {
...
+                <Button
+                  size="small"
+                  to={`/settings/${orgId}/integrations/${provider.key}/${integration.id}/`}
+                >
+                  {t('Configure')}
+                </Button>
+              </Box>
+            )}
+          {integration.status === 'disabled' &&
+            this.hasConfiguration() && (
```
Comment (EvanPurkhiser) [inline src/sentry/static/sentry/app/views/organizationIntegrations/installedIntegration.jsx]: This doesn't seem right, the reinstall button would only show up if it's configurable because of the `hasConfiguration`?

**Review by evanpurkhiser — CHANGES_REQUESTED** [review]

File: src/sentry/integrations/pipeline.py (hunk 1 above)
Comment (EvanPurkhiser) [inline src/sentry/integrations/pipeline.py]: Just realized by early returning we're also losing the `self.clear_session`

File: src/sentry/api/serializers/models/integration.py [hunk 12]
```diff
@@ -13,16 +13,19 @@ def serialize(self, obj, attrs, user):
         provider = obj.get_provider()
         return {
             'id': six.text_type(obj.id),
+            'external_id': obj.external_id,
```
Comment (EvanPurkhiser) [inline src/sentry/api/serializers/models/integration.py]: Ooop, this should be `externalId`

Comment (EvanPurkhiser) [reply src/sentry/api/serializers/models/integration.py]: Pointed this out in standup, lets just use the `id` instead of external ID to identify what integration we're reinstalling

File: src/sentry/integrations/pipeline.py [hunk 13]
```diff
@@ -55,7 +55,14 @@ def finish_pipeline(self):
         return response
 
     def _finish_pipeline(self, data):
-        if 'expect_exists' in data:
+        if 'reinstall_id' in data:
+            integration = Integration.objects.get(
+                provider=self.provider.key,
+                external_id=data['reinstall_id'],
+            )
+            integration.reinstall(data['external_id'])
```
Comment (EvanPurkhiser) [inline src/sentry/integrations/pipeline.py]: Actually I was thinking the `reinstall` method could live on the integration service object. So instead of the model knowing about reinstall logic, the integration would know what it needs to do to reinstall itself.

```
self.update(external_id=data['external_id'], status=ObjectStatus.VISIBLE)
integration.get_intsallation().reinstall()
```

Note that the updating of the external ID just happens right here. Then the repository updating would actually happen in the github specific integration stuff (or ideally, base mixin class for repository functionality of integrations)

File: src/sentry/static/sentry/app/views/organizationIntegrations/addIntegrationButton.jsx [hunk 14]
```diff
@@ -48,12 +50,16 @@ export default class AddIntegrationButton extends React.Component {
 
   handleAddIntegration = provider => {
     const name = 'sentryAddIntegration';
+    debugger;
```
Comment (EvanPurkhiser) [inline src/sentry/static/sentry/app/views/organizationIntegrations/addIntegrationButton.jsx]: oops

Comment (billyvg) [reply src/sentry/static/sentry/app/views/organizationIntegrations/addIntegrationButton.jsx]: how did this pass precommit hooks :<

File: src/sentry/static/sentry/app/views/organizationIntegrations/addIntegrationButton.jsx [hunk 15]
```diff
@@ -85,21 +91,21 @@ export default class AddIntegrationButton extends React.Component {
...
+    const {provider, onAddIntegration, reinstall, ...buttonProps} = this.props;
+    let buttonText = reinstall ? t('Reinstall') : t('Add') + ' ' + provider.metadata.noun;
 
     return (
       <Tooltip
         disabled={provider.canAdd}
         tooltipOptions={{placement: 'left'}}
-        title={`Integration cannot be added on Sentry. Enable this integration via the ${provider.name} instance.`}
-      >
+        title={`Integration cannot be added on Sentry. Enable this integration via the ${provider.name} instance.`}>
```
Comment (EvanPurkhiser) [inline src/sentry/static/sentry/app/views/organizationIntegrations/addIntegrationButton.jsx]: This is weird, are you using a different version of prettier. Afaik the formatting of multiline jsx elements has the close tag on a new line /cc @billyvg

File: src/sentry/static/sentry/app/views/organizationIntegrations/installedIntegration.jsx [hunk 16]
```diff
@@ -16,7 +18,7 @@ export default class InstalledIntegration extends React.Component {
     provider: PropTypes.object.isRequired,
     integration: PropTypes.object.isRequired,
     onRemove: PropTypes.func.isRequired,
-    onDisable: PropTypes.func.isRequired,
+    onDisable: PropTypes.func.isRequired
```
Comment (EvanPurkhiser) [inline src/sentry/static/sentry/app/views/organizationIntegrations/installedIntegration.jsx]: Why'd the comma get removed :(

File: src/sentry/static/sentry/app/views/organizationIntegrations/addIntegrationButton.jsx (hunk 14 above)
Comment (EvanPurkhiser) [reply src/sentry/static/sentry/app/views/organizationIntegrations/addIntegrationButton.jsx]: https://travis-ci.org/getsentry/sentry/jobs/397900763#L655 It actually only shows up as a warning

File: src/sentry/static/sentry/app/views/organizationIntegrations/installedIntegration.jsx [hunk 17]
```diff
@@ -77,20 +95,34 @@ export default class InstalledIntegration extends React.Component {
...
             </Box>
           )}
+          {integration.status === 'disabled' && (
+            <Box mr={1}>
+              <AddIntegrationButton
+                size="small"
+                priority="danger"
+                provider={provider}
+                integration={integration}
+                onAddIntegration={this.mergeIntegration}
```
Comment (EvanPurkhiser) [inline src/sentry/static/sentry/app/views/organizationIntegrations/installedIntegration.jsx]: Yeah this should call something like `this.props.reinstallIntegration` and then that will be set up a component to `mergeIntegration`

**Review by evanpurkhiser — APPROVED** [review]
